### PR TITLE
Subclass Sync - Direct in source, indirect in Mondo

### LINKED ISSUE #708: Bug in Subclass Sync: Missing evidence in "confirmed" table
See comment here: https://github.com/monarch-initiative/mondo/pull/8432#issuecomment-2493386699

## _Bug_: 

There are 

- pairs of Mondo classes with a direct subclass relation
- both exact matches to DOID
- ...but the subclass relation _does not appear in the [list of confirmed subclasses](https://raw.githubusercontent.com/monarch-initiative/mondo-ingest/main/src/ontology/reports/sync-subClassOf.confirmed.tsv) (`reports/sync-subClassOf.confirmed.tsv`)_.

## Example:

```
[Term]
id: MONDO:0000248
name: dengue shock syndrome
xref: DOID:0050125 {source="MONDO:equivalentTo"}
is_a: MONDO:0005502 {source="DOID:0050125", source="MONDO:Redundant"} ! dengue disease
```

```
[Term]
id: MONDO:0005502
name: dengue disease
xref: DOID:12205 {source="MONDO:equivalentTo", source="EFO:0005547"}
```

And in DO:

```
<owl:Class rdf:about="http://purl.obolibrary.org/obo/DOID_0050125">
    <rdfs:subClassOf rdf:resource="http://purl.obolibrary.org/obo/DOID_12205"/>
    <rdfs:label xml:lang="en">dengue shock syndrome</rdfs:label>
</owl:Class>
```

The confirmed subclass table is clearly missing this, only containing these two entries related to dengue:

```
MONDO:0005358	Dengue hemorrhagic fever	MONDO:0005502	DOID:12206	DOID:12205	dengue disease
MONDO:0000259	asymptomatic dengue	MONDO:0005502	DOID:0050143	DOID:12205	dengue disease
```

In my opinion its a bug in the subclass sync, one that we didn't notice because my previous subclass pipeline didn't delete _all_ the evidence.

### PR DESCRIPTION
**Pull request by joeflack4**
Resolves #708

## Overview
Added code to implement Subclass Sync "[case 2](https://docs.google.com/document/d/1H8fJKiKD-L1tfS-2LJu8t0_2YXJ1PQJ_7Zkoj7Vf7xA/edit?tab=t.0#heading=h.9hixairfgxa1)": Direct in source, indirect in Mondo.

### Documentation
Was the documentation added/updated under `docs/`?
- [ ] Yes
- [x] No, updates to the docs were not necessary after careful consideration

### QC
Was the full pipeline run before submitting this PR using `sh run.sh make build-mondo-ingest` on this branch (after 
`docker pull obolibrary/odkfull:dev`), and no errors occurred?
- [x] Yes
- [ ] No, there are no functional (code-related) changes to the pipeline in the PR, so no re-run is necessary
   
Mini build:
- #717

### New Packages
Were any new *Python packages* added?
- [ ] Yes, [requirements.txt.full](https://github.com/INCATools/ontology-development-kit/blob/master/requirements.txt.full) was updated and an [ODK issue](https://github.com/INCATools/ontology-development-kit/issues/new) submitted
- [x] No

Were any other *non-Python packages* added?
- [ ] Yes, [Dockerfile](https://github.com/INCATools/ontology-development-kit/blob/master/Dockerfile) updated and an [ODK issue](https://github.com/INCATools/ontology-development-kit/issues/new) submitted
- [x] No

### PR Review and Conversations Resolved
Has the PR been sufficiently reviewed by at least 1 team member of the Mondo Technical team and all threads resolved?
- [x] Yes

## Additional info
I ran output and observed that the case in #708 appeared as expected!:

|subject_mondo_id|subject_mondo_label|object_mondo_id|subject_source_id|object_source_id|object_mondo_label|
|-|-|-|-|-|-|
|ID| |SC %|>A oboInOwl:source| | |
|MONDO:0000248|dengue shock syndrome|MONDO:0005502|DOID:0050125|DOID:12205|dengue disease|

See: Google Sheets:
- [DO](https://docs.google.com/spreadsheets/d/1FUoYJPASTCps55PUCEUtspdY7e5CgSSv245Z_O5SxOk/edit?gid=21333276#gid=21333276)
- [`sync-subClassOf.confirmed-direct-source-indirect-mondo.tsv`](https://docs.google.com/spreadsheets/d/1rAWuzumhr4XEgGOhACsqUQfk3-JiWp8CpTVkbwgProg/edit?gid=1987521551#gid=1987521551)

### CODE REVIEW

File: src/ontology/mondo-ingest.Makefile [hunk 1]
```diff
@@ -567,14 +567,18 @@ $(TMPDIR)/sync-subClassOf.added.self-parentage.tsv: $(foreach n,$(ALL_COMPONENT_
 $(REPORTDIR)/sync-subClassOf.direct-in-mondo-only.tsv: $(foreach n,$(ALL_COMPONENT_IDS), $(REPORTDIR)/$(n).subclass.direct-in-mondo-only.tsv) $(TMPDIR)/mondo.db
 	python3 $(SCRIPTSDIR)/sync_subclassof_collate_direct_in_mondo_only.py --outpath $@
 
-$(REPORTDIR)/sync-subClassOf.confirmed.tsv: $(foreach n,$(ALL_COMPONENT_IDS), $(REPORTDIR)/$(n).subclass.confirmed.robot.tsv)
+$(REPORTDIR)/sync-subClassOf.confirmed.tsv: $(foreach n,$(ALL_COMPONENT_IDS), $(REPORTDIR)/$(n).subclass.confirmed.robot.tsv) $(TMPDIR)/mondo.db
```
Comment (matentzn) [inline src/ontology/mondo-ingest.Makefile:570]: Why is $(TMPDIR)/mondo.db added specifically to this goal as a dependency?

File: src/ontology/mondo-ingest.Makefile [hunk 2]
```diff
@@ -567,14 +567,18 @@ $(TMPDIR)/sync-subClassOf.added.self-parentage.tsv: $(foreach n,$(ALL_COMPONENT_
...
 	awk '(NR == 1) || (NR == 2) || (FNR > 2)' $(REPORTDIR)/*.subclass.confirmed.robot.tsv > $@
 
-$(REPORTDIR)/%.subclass.confirmed.robot.tsv $(REPORTDIR)/%.subclass.added.robot.tsv $(REPORTDIR)/%.subclass.added-obsolete.robot.tsv $(REPORTDIR)/%.subclass.direct-in-mondo-only.tsv $(TMPDIR)/%.subclass.self-parentage.tsv: $(TMPDIR)/mondo-ingest.db $(TMPDIR)/mondo.db $(TMPDIR)/mondo.sssom.tsv
+$(REPORTDIR)/sync-subClassOf.confirmed-direct-source-indirect-mondo.tsv: $(foreach n,$(ALL_COMPONENT_IDS), $(REPORTDIR)/$(n).subclass.confirmed-direct-source-indirect-mondo.robot.tsv) $(TMPDIR)/mondo.db
+	awk '(NR == 1) || (NR == 2) || (FNR > 2)' $(REPORTDIR)/*.subclass.confirmed-direct-source-indirect-mondo.robot.tsv > $@
```
Comment (matentzn) [inline src/ontology/mondo-ingest.Makefile:574]: Wow I have never seen this pattern `awk '(NR == 1) || (NR == 2) || (FNR > 2)'`! Super useful!

**Review by matentzn — COMMENTED** [review]
I kinda confirm it after scrolling through, but what matters is the output PR.

File: src/ontology/mondo-ingest.Makefile (hunk 2 above)
Comment (joeflack4) [reply src/ontology/mondo-ingest.Makefile:574]: You can thank Claude! Fortunately one of the few succinct `awk` patterns it's shown me.

File: src/ontology/mondo-ingest.Makefile (hunk 1 above)
Comment (joeflack4) [reply src/ontology/mondo-ingest.Makefile]: Thanks for noticing! This is a mistake. This (and another goal, too) do not need this prereq. Removed~

**Review by twhetzel — COMMENTED** [review]
I provisionally accept pending discussion in https://github.com/monarch-initiative/mondo-ingest/pull/717

Update: After looking over #717 further, it is correctly identifying the Case "SCR is direct in source, but indirect Mondo" so I will approve.

**Review by twhetzel — APPROVED** [review]

**Review by matentzn — CHANGES_REQUESTED** [review]
given @twhetzel observations:

Please do _not_ include subclass links to Mondo:disease or Mondo:HumanDisease in the output!

**Review by twhetzel — CHANGES_REQUESTED** [review]
Updated review to be consistent with the new request to
1.  _not_ include subclass links to Mondo:disease or Mondo:HumanDisease in the output
2. get rid of cases already reflected by direct to direct file (see [this comment](https://github.com/monarch-initiative/mondo-ingest/pull/717#discussion_r1882897370))

@joeflack4 are these requests clear and can you get this wrapped up tomorrow?

File: src/scripts/sync_subclassof.py [hunk 3]
```diff
@@ -298,10 +316,12 @@ def sync_subclassof(
     # - rels_indirect_source_source: Indirect relationships (source IDs) from source
     # - rels_indirect_source_mondo:  Relationships from rels_indirect_source_source that appear in Mondo
     # - rels_indirect_source_mondo_and_1or2_ids_not_in_mondo: from rels_indirect_source_source not in Mondo (unused)
+    # - rels_indirect_mondo_source:  Relationships from rels_indirect_mondo_mondo that appear in source
     logging.info('- Collecting indirect subclass relations')
     rels_indirect_mondo_mondo = ancestors_mondo_mondo.difference(rels_direct_mondo_mondo)
-    # todo: remove unused, commented out vars?
-    # rels_indirect_mondo_source = ''  # not needed?
+    rels_indirect_mondo_source, rels_indirect_mondo_mondo_and_1or2_ids_not_in_source = _convert_edge_namespace(
```
[the remark this answers is not in the document]
Comment (joeflack4) [reply src/scripts/sync_subclassof.py:322]: I hate these variable names. I mentioned in a recent call w/ Nico that this pipeline is in need of OO. I think that would make it easier to understand and resolve the issue of these weird variable names.

It is a long, descriptive variable name though, and I do have some documentation comments about these variable names, though I don't think there is a comment for `rels_indirect_mondo_mondo_and_1or2_ids_not_in_source` (and there are 1-2 similar variables, I think).

The `_1or2_` actually is meant to be quite descriptive / accurate. Perhaps I can add another comment: It means that for a given subclass pair (parent, child), either 1 of the terms (parent or child) is not in the source, or both of them (2) are not in the source.

[the remark this answers is not in the document]
Comment (twhetzel) [reply src/scripts/sync_subclassof.py:322]: yeah, i've seen other code where you've used this naming convention but what the numbers referred to was not obvious without some digging so at a very, very bare minimum, if these types of variable names are used at least add a comment (which should be brief) to describe what the variable is actually referring to if you're able to think of a better variable name at the time you are writing the code. This is only for the future, I am not asking that you add comments here in this code or for any other variables that are like this right now.

**Comment by joeflack4** [comment]
@twhetzel Yes, both tasks are clear! I will try my best to complete them tomorrow. If not, it will be done by Monday for sure.

RE: `MONDO:'human disease'` (`MONDO:0700096`), I see 0 usages of this in the output currently but I will ensure that these will be filtered out anyway should they appear! I do confirm the same `n` that you found for occurrences of 'disease' (`MONDO:0000001`).

**Comment by joeflack4** [comment]
@twhetzel OK, I'm done with these updates! See: ffc0b1580f6f7463c32304a6181c1a6961792603

Updated the [google sheet](https://docs.google.com/spreadsheets/d/1rAWuzumhr4XEgGOhACsqUQfk3-JiWp8CpTVkbwgProg/edit?gid=1987521551#gid=1987521551).

More details in the [mini-build thread](https://github.com/monarch-initiative/mondo-ingest/pull/717#discussion_r1882897370).

**Review by matentzn — APPROVED** [review]
LGTM !

**Review by twhetzel — APPROVED** [review]
LGTM - were there updated files to review?

**Pull request merged**